outlines/fsm/json_schema.py

```diff
@@ -102,13 +100,43 @@ def to_regex(resolver: Resolver, instance: dict):
     if "properties" in instance:
         regex = ""
         regex += r"\{"
-        for i, (name, value) in enumerate(instance["properties"].items()):
-            regex += f'{whitespace}"{name}"{whitespace}:{whitespace}'
-            regex += to_regex(resolver, value)
-
-            # No comma after the last key-value pair in JSON
-            if i < len(instance["properties"]) - 1:
-                regex += f"{whitespace},"
+        properties = instance["properties"]
+        required_properties = instance.get("required", [])
+        is_required = [item in required_properties for item in properties]
+        # If at least one property is required, we include the one in the lastest position
```

Comment:
Is this changing the order of the fields as they were specified in the JSON schema?

Reply:
```python
We need to make sure that the last item generated does not end with a comma and thus end up with invalid JSON.

When at least one property is required, we select the last in the list and include it at the last position without a comma. We then prepend every prior (optional or required) property with a comma.
```

Reply:
Why not do the same as below, actually?

Reply:
This is not changing the order of the fields.
An issue with `When at least one property is required, we select the last in the list and include it at the last position without a comma` is that it would change the order of the fields if the last property is an optional one.

Reply:
I can't see a way of easily combining both as the case in which there are no required properties creates as many subregexes as there are properties and connect them with `|` while the case with at least one required property does not need that.